crates/chain/src/indexed_tx_graph.rs

```diff
@@ -0,0 +1,413 @@
+use core::convert::Infallible;
+
+use bitcoin::{OutPoint, Transaction, TxOut};
+
+use crate::{
+    keychain::Balance,
+    sparse_chain::ChainPosition,
+    tx_graph::{Additions, TxGraph, TxInGraph},
+    BlockAnchor, ChainOracle, FullTxOut, ObservedIn, TxIndex, TxIndexAdditions,
+};
+
+/// An outwards-facing view of a transaction that is part of the *best chain*'s history.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TxInChain<'a, T, A> {
+    /// Where the transaction is observed (in a block or in mempool).
+    pub observed_in: ObservedIn<&'a A>,
+    /// The transaction with anchors and last seen timestamp.
+    pub tx: TxInGraph<'a, T, A>,
+}
+
+/// An outwards-facing view of a relevant txout that is part of the *best chain*'s history.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TxOutInChain<'a, I, A> {
+    /// The custom index of the txout's script pubkey.
+    pub spk_index: &'a I,
+    /// The full txout.
+    pub txout: FullTxOut<ObservedIn<&'a A>>,
+}
+
+/// A structure that represents changes to an [`IndexedTxGraph`].
+#[derive(Clone, Debug, PartialEq)]
+#[must_use]
+pub struct IndexedAdditions<A, D> {
+    /// [`TxGraph`] additions.
+    pub graph_additions: Additions<A>,
+    /// [`TxIndex`] additions.
+    pub index_delta: D,
+    /// Last block height witnessed (if any).
+    pub last_height: Option<u32>,
+}
+
+impl<A, D: Default> Default for IndexedAdditions<A, D> {
+    fn default() -> Self {
+        Self {
+            graph_additions: Default::default(),
+            index_delta: Default::default(),
+            last_height: None,
+        }
+    }
+}
+
+impl<A: BlockAnchor, D: TxIndexAdditions> TxIndexAdditions for IndexedAdditions<A, D> {
```

Comment:
why is `TxIndexAdditions` implemented for `IndexAdditions`???

I think this trait doesn't need to exist at all.

Reply:
Don't we need a way for the `Additions` of the `TxIndex` trait to be able to append on top of one another?

Otherwise, how do we ensure `IndexedAdditions` can append on top of one another?

Reply:
> Don't we need a way for the `Additions` of the `TxIndex` trait to be able to append on top of one another?
> 
> Otherwise, how do we ensure `IndexedAdditions` can append on top of one another?

Where do we need this/why do we want to ensure this?

Reply:
Removed the `TxIndexAdditions` trait: 90525ec799d669a7ee39ada3be258a06cb6adb49